### PR TITLE
Multipart upload

### LINKED ISSUE #2691: Support multipart file upload in storage api
## Description

The storage api currently supports `PUT`ing a file as binary data, which is great for machine clients, but to be used from a browser, it will need to support [Form Multipart](https://stackoverflow.com/questions/16958448/what-is-http-multipart-request) as a source of uploads.

## Acceptance Criteria
-   It's possible to upload to the storage api from a super-simple browser form

## Technical Notes
- This probably means using express multiparty: https://stackoverflow.com/questions/23027737/multipart-form-data-upload-nodejs-expressjs

## Out of Scope
- Making a proper UI that uses this - this issue is purely focused on providing the backend capability to do an upload from a browser

### PR DESCRIPTION
**Pull request by sajidanower23**
### What this PR does

Fixes #2691 

Adds a new `POST /upload/:bucket` endpoint for multipart uploads (particular from the browser)

### Checklist

-   [x] There are unit tests to verify my changes are correct
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: magda-storage-api/src/createApiRouter.ts [hunk 1]
```diff
@@ -144,6 +145,69 @@ export default function createApiRouter(options: ApiRouterOptions) {
...
+     * {
+     *      "message": "Successfully uploaded 2 files.",
+     *      "etags": ["cafbab71cd98120b777799598f0d4808-1","19a3cb5d5706549c2f1a57a27cf30e41-1"]}
+     *
+     * @apiErrorExample {string} 500
+     *      Internal server error.
+     */
+    router.post(
+        "/upload/:bucket",
+        fileParser({ rawBodyOptions: { limit: "10mb" } }),
```
Comment (AlexGilleran) [inline magda-storage-api/src/createApiRouter.ts]: Might need a higher (configurable) limit for our purposes

**Comment by sajidanower23** [comment]
Made a deployment to https://2691-multipart-upload.dev.magda.io. To test:
```console
$ curl -X POST --cookie 'connect.sid=<cookie>' -F 'data=@/path/to/file' https://2691-multipart-upload.dev.magda.io/api/v0/storage/upload/magda-bucket
{"message":"Successfully uploaded 1 file(s).","etags":["39c590948abc8a688970fbcf86e756bc-1"]}
```

File: magda-storage-api/src/createApiRouter.ts (hunk 1 above)
Comment (sajidanower23) [reply magda-storage-api/src/createApiRouter.ts]: Thanks for picking that up. Upload limit should be configurable now :+1:

File: magda-storage-api/src/test/storage.spec.ts [hunk 2]
```diff
@@ -175,6 +177,34 @@ describe("Storage API tests", () => {
...
+                    authApiUrl,
+                    true,
+                    jwtSecret,
+                    request(app)
+                        .post("/v0/upload/" + bucketName)
+                        .field("originalname", "test-browser-upload-1")
+                        .expect(400)
+                );
+            });
+        });
```
Comment (AlexGilleran) [inline magda-storage-api/src/test/storage.spec.ts]: Could use a test to ensure that authorisation is required too :)

File: magda-storage-api/src/MagdaMinioClient.ts [hunk 3]
```diff
@@ -151,7 +151,7 @@ export default class MagdaMinioClient implements ObjectStoreClient {
             return this.client.removeObject(bucket, objectName, function(
                 err: any
             ) {
-                console.log(err);
+                console.error(err);
```
[the remark this answers is not in the document]
Comment (sajidanower23) [reply magda-storage-api/src/MagdaMinioClient.ts]: Removed

File: magda-storage-api/src/test/storage.spec.ts [hunk 4]
```diff
@@ -175,6 +177,34 @@ describe("Storage API tests", () => {
...
+                return mockAuthorization(
+                    authApiUrl,
+                    true,
+                    jwtSecret,
+                    request(app)
+                        .post("/v0/upload/" + bucketName)
+                        .field("originalname", "test-browser-upload-1")
+                        .attach("image", "src/test/test_image.jpg")
+                        .attach("text", "src/test/test_csv_1.csv")
+                        .expect(200)
```
[the remark this answers is not in the document]
Comment (sajidanower23) [reply magda-storage-api/src/test/storage.spec.ts]: Done, finally

File: magda-storage-api/src/test/storage.spec.ts (hunk 2 above)
Comment (sajidanower23) [reply magda-storage-api/src/test/storage.spec.ts]: Added

File: magda-storage-api/src/createApiRouter.ts [hunk 5]
```diff
@@ -174,13 +174,12 @@ export default function createApiRouter(options: ApiRouterOptions) {
...
-            const contentType = req.headers["content-type"];
             const contentLength = req.headers["content-length"];
-            const metaData = {
-                "Content-Type": contentType,
-                "Content-Length": contentLength
-            };
             const promises = (req.files as Array<any>).map((file: any) => {
+                const metaData = {
+                    "Content-Type": file.mimetype,
+                    "Content-Length": contentLength,
```
[the remark this answers is not in the document]
Comment (sajidanower23) [reply magda-storage-api/src/createApiRouter.ts]: Good catch! Content-Length can still be recorded by getting the byteLength of the file buffer, which I just did :+1:

[the remark this answers is not in the document]
Comment (AlexGilleran) [reply magda-storage-api/src/createApiRouter.ts]: 🎉

**Pull request merged**